capstone/src/test/java/com/google/sps/servlets/UserDataServletTest.java

```diff
@@ -0,0 +1,138 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Arrays;
+import java.util.ArrayList;
+
+/**
+ * Unit tests for {@link UserDataServlet}.
+ */
+ @RunWith(JUnit4.class)
+public class UserDataServletTest {
+  private static final String USER_EMAIL = "test@mctest.com";
+  private static final String USER_ID = "testy-mc-test";
+
+  // Set no eventual consistency, that way queries return all results.
+  // https://cloud.google.com/appengine/docs/java/tools/localunittesting
+  // #Java_Writing_High_Replication_Datastore_tests
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+            new LocalDatastoreServiceTestConfig()
+                .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+            new LocalUserServiceTestConfig())
+          .setEnvEmail(USER_EMAIL)
+          .setEnvIsLoggedIn(true)
+          .setEnvAuthDomain("gmail.com")
+          .setEnvAttributes(
+              new HashMap(
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  private static final ArrayList<String> INTERESTS_LIST = new ArrayList<String>( 
+      Arrays.asList("Testing", "Dancing"));
+  private static final User USER_1 = new User(USER_ID, "Test", "McTest", USER_EMAIL, 
+                          /* phoneNumber= */ "123-456-7890", 
+                          /* profilePic= */ "", 
+                          /* badges= */ new LinkedHashSet<Badge>(), 
+                          /* groups= */ new LinkedHashSet<Long>(), 
+                          /* interests= */ INTERESTS_LIST);
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private UserDataServlet userDataServlet;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    populateDatabase(datastore);
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    userDataServlet = new UserDataServlet();
+  }
+
+  @After
+  public void tearDown() {
+    helper.setEnvIsLoggedIn(true);
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGet_retrieveUserData() throws Exception {
+    userDataServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+
+    String expectedResponse = new Gson().toJson(USER_1);
+
+    // remove any whitespace from JSON
+    response = response.replaceAll("\\s", "");
+    expectedResponse = expectedResponse.replaceAll("\\s", "");
```

Comment:
I'm curious, are the json object generating different whitespace? It may be helpful to remove this whitespace stripping logic into a helper function, maybe something like `assertEqualJson`. I really don't like Util files, but I do see this is a problem in multiple test files, so it could be something like a `TestUtils` file.

Reply:
Yeah, not sure why but I think in the response there's sometimes an extra newline or carriage return that causes the equality tests to fail. I'll put it in a util file